tsl/src/continuous_aggs/materialize.c

```diff
@@ -660,6 +674,66 @@ scan_take_invalidation_tuple(TupleInfo *ti, void *data)
 	return SCAN_CONTINUE;
 }
 
+static ScanTupleResult
+invalidation_threshold_htid_found(TupleInfo *tinfo, void *data)
+{
+	if (tinfo->lockresult != TM_Ok)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("unexpected lockresult %d in invalidation_threshold_htid_found",
+						tinfo->lockresult)));
+	}
+	return SCAN_DONE;
+}
+
+/* lock row corresponding to hypertable id in
+ * continuous_aggs_invalidation_threshold table in AccessExclusive mode
+ * block till lock is acquired.
+ */
+static void
+lock_invalidation_threshold_hypertable_row(int32 raw_hypertable_id)
+{
+	ScanTupLock scantuplock = {
+		.waitpolicy = LockWaitBlock,
+		.lockmode = LockTupleExclusive,
+	};
+	Relation invalidation_threshold_table_relation;
+	Catalog *catalog = ts_catalog_get();
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_continuous_aggs_invalidation_threshold_pkey_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(raw_hypertable_id));
+
+	/* lock table in RowExclusive mode and the row with AccessExclusive */
+	ScannerCtx scanctx = { .table = catalog_get_table_id(catalog,
+														 CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+						   .index = catalog_get_index(catalog,
+													  CONTINUOUS_AGGS_INVALIDATION_THRESHOLD,
+													  CONTINUOUS_AGGS_INVALIDATION_THRESHOLD_PKEY),
+						   .nkeys = 1,
+						   .scankey = scankey,
+						   .limit = 1,
+						   .tuple_found = invalidation_threshold_htid_found,
+						   .lockmode = RowExclusiveLock,
+						   .scandirection = ForwardScanDirection,
+						   .result_mctx = CurrentMemoryContext,
+						   .tuplock = &scantuplock };
+	/* acquire lock explicitly.
+	 * the catalog scanning code , releases it at end of the call
+	 */
+	invalidation_threshold_table_relation =
+		table_open(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+				   RowExclusiveLock);
+	ts_scanner_scan(&scanctx);
+	table_close(invalidation_threshold_table_relation, NoLock); /*release lock at end
```

Comment:
I don't understand why we keep the table lock here (in `RowExclusiveLock`) mode. That seems to negate the purpose of taking the tuple lock since you are now serializing around the invalidation threshold table instead of the tuple. So, in effect, this is a much more heavy-handed (global) serialization that blocks refreshes across all caggs for all hypertables.

Or, am I missing something here?

Reply:
changed to AccessShare  based on our discssion.

Reply:
Lock still held here. I thought that the table lock was going to be released here since tuple lock is held and that is the actual serialization point.

Reply:
I misunderstood that - modified so that we only have the tuple lock.